<a href="https://colab.research.google.com/github/weymouth/NumericalPython/blob/main/02ConditionalsAndLists.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Conditionals and Lists in Python

We now have the basic building blocks of variables, operations and functions in Python, allowing us to perform a wide range of calculations. However, adding conditionals and lists to our Python capabilities will greatly increase the range of engineering problems we can solve.

![Towing tank](https://cdn.southampton.ac.uk/assets/imported/transforms/content-block/CB_RImg/19677430F8FD415F823DAC20DFE72BBE/DSCF0072-banner.jpg_SIA_JPG_background_image.jpg)

As an example to guide this notebook, we will determining wave conditions in an experimental tank such as the Bolderwood towing tank in the image above.

# Booleans and Conditionals

You can test the value of a variable and use this to control the execution of a program using conditional statements. 

## Booleans

The values `True` and `False` are built-in Boolean objects. We can build up logical conditions and tests and the result will be one of these objects. Predict what the logical statements below will print. Then test your predictions by using the "play" button or pressing [shift]-[enter].

In [33]:
a = True
b = True
c = False
print(type(a))
print( not a )
print( a and b )
print( a and c )
print( a or c )
print( not a or (c or b) )

<class 'bool'>
False
True
False
True
True


In numerical programming and analysis, Booleans typically occur when testing the relationship between two numbers using boolean operators

| Operation     | Description                       || Operation     | Description                          |
|---------------|-----------------------------------||---------------|--------------------------------------|
| ``a == b``    | ``a`` equal to ``b``              || ``a != b``    | ``a`` not equal to ``b``             |
| ``a < b``     | ``a`` less than ``b``             || ``a > b``     | ``a`` greater than ``b``             |
| ``a <= b``    | ``a`` less than or equal to ``b`` || ``a >= b``    | ``a`` greater than or equal to ``b`` |

Like before, try to guess what these condional operations will return before running the code block below:

In [34]:
print( 5 == 5 )
print( 5. != 5 )
print( 'five' == 5 )
print( 10/2 != 25**(.5) )

print( 4 >= 4 )
print( 4+1/1000000 > 4 )
print( (4+1)/10000 >= 4 )

print( 4>=5 or 5>=4 )

True
False
False
False
True
True
False
True


And, not surprisingly, you can also write functions to evaluate boolean checks. A classic example is to check if a number is even or odd using the mod operator:

In [35]:
def is_odd(i):
    return i%2==1

for i in range(4):
    print("The number {} is odd: {}".format(i,is_odd(i)))

The number 0 is odd: False
The number 1 is odd: True
The number 2 is odd: False
The number 3 is odd: True


As you can see, this correctly identifies even and odd digits.

As usual, I'm sneaking in a few more ideas into this example. In this case, I've used a [format statement](https://www.w3schools.com/python/ref_string_format.asp) to insert the number we were checking `i` and the output of the function `is_odd(i)` into a string and then print it. This is just "for looks" in this case, but it is a nice technique to know.

## Conditional statements

One of the more powerful aspects of numerical programming is derived by combining booleans with conditional statements like `if`. The following example shows the syntax for a python `if` statement.

In [36]:
def describe_number(n):
    if is_odd(n):
        print(n,' is odd.')
    elif n>0:
        print(n,' is even and greater than 0.')
    else:
        print(n,' is even and not greater than 0.')

for n in range(-2,3):
    describe_number(n)

-2  is even and not greater than 0.
-1  is odd.
0  is even and not greater than 0.
1  is odd.
2  is even and greater than 0.


This function uses a different print statement based on the properties of the number. In numerical computing we often use these code "branches" to choose different formulas depending on the input. 

**Example:** The towing take has a wave maker and we need to know how long those waves will be $\lambda$ as a function of the frequency of the wave maker paddle $\omega$ and the depth of the tank $h$. There is a simple formula for this, but it only holds in deep water,

$$ \lambda = \frac{2\pi g}{\omega^2} \quad \text{if}\quad \lambda < 2h $$

where $g$ is the acceleration of gravity. If the wave is too long, we need to use a more general form of the dispersion relationship which we will look at later. 

So let's write a function for this problem

In [37]:
def deep_wavelength(omega):
    # omega needs to be given in rad/s.
    g = 9.81 # m/s**2
    pi = 3.14159
    return 2*pi*g/omega**2 # length in m

def check_wavelength(omega,h=3):
    # omega needs to be given in rad/s and h in m.
    l = deep_wavelength(omega)
    if l<2*h:
        print("Wavelength is {:.3g} m when omega = {} rad/s".format(l,omega))
    else:
        print('Wave is not deep water ' 
              'when h = {} m and omega = {} rad/s.'.format(h,omega))
        
for omega in range(2,8):
    check_wavelength(omega)

Wave is not deep water when h = 3 m and omega = 2 rad/s.
Wave is not deep water when h = 3 m and omega = 3 rad/s.
Wavelength is 3.85 m when omega = 4 rad/s
Wavelength is 2.47 m when omega = 5 rad/s
Wavelength is 1.71 m when omega = 6 rad/s
Wavelength is 1.26 m when omega = 7 rad/s


In the first function computes the deep water wavelength and the second either prints that length or that it isn't valid. 

Note that I have set `h=3` in the first line (the _header_) of `check_wavelength`. This makes `h` an _optional argument_ with a default value of 3 (since our wavetank is $3 \text{m}$ deep). So `check_wavelength(10)==check_wavelength(10,3)`. In the function body I compute the value of wavelength by calling the deep water function. Then I check to make sure that formula was valid, and print the appropriate statement using `if-else`. The only new syntax here is in the last print statement: `{:.3g}` converts the float `l` to a string with 3 significant digits.

# Lists

A collection of variables can be grouped together in many types of [data structures](https://docs.python.org/3/tutorial/datastructures.html) in Python. In this notebook we focus on lists.

A list in python is just sequence of variables. We've had many examples now that use `range()` to make a list of integers, but there are many more choices. For example

In [38]:
primes = [2,3,5,7,11,13,17]
primes

[2, 3, 5, 7, 11, 13, 17]

In [39]:
single_name_singers = ["Prince","P!nk","Eminem","Rhianna","Madonna","Beyonc\u00E9"]
single_name_singers

['Prince', 'P!nk', 'Eminem', 'Rhianna', 'Madonna', 'Beyoncé']

(Note the use of unicode to get the _grave_ in Beyoncé.)

As you can see, the format for a list is to enclose the items in square brackets `[ ]` and to separate each item with a comma `,`. If nothing goes inside, you just get an empty list.

In [40]:
empty = []
empty

[]

And you can loop through the items in any list just as we've done before using `for-in`

In [41]:
for name in single_name_singers:
    print(name)

Prince
P!nk
Eminem
Rhianna
Madonna
Beyoncé


In [42]:
household_ages = [1/3,2,4,9,11] # I have a full house
for p in primes:
    if p in household_ages: 
        print(p,' is prime AND the age of one of my children/pets')

2  is prime AND the age of one of my children/pets
11  is prime AND the age of one of my children/pets


## Sublists

But we don't need to go through all the items in a list. We can also _index_ a particular item we want, or loop through a _slice_ of a list. 

Python uses 0-based indexing so the first item is has index 0, the second has index 1 and so on.

In [43]:
print(primes[0])
print(primes[1])

2
3


So what is the index of the **last** item in our list of primes? Modify the code above to check your math. 

You can also count backwards through a list. Guess what the results before you run this cell:

In [44]:
print(primes[-1])
print(primes[-2])

17
13


Finally, we can select only part of an array, a *slice*. Similar to the syntax of the `range` function, you can index using  `[start]:span[:step]`, where the items in the square brackets are optional. For example, who are the first three singers in the list? 

In [45]:
single_name_singers[:3]

['Prince', 'P!nk', 'Eminem']

Try some more advanced slicing on your own to get the sublists suggested below.

In [46]:
# 1. Slice to give the last three singers
print(single_name_singers[3:])
# 2. Slice to give every other prime, starting at 3. ie 3,7,11,...
primes[1:]

['Rhianna', 'Madonna', 'Beyoncé']


[3, 5, 7, 11, 13, 17]

## Functions and methods

There are many potentially helpful functions that can applied to lists. I've summarized a few below:



In [47]:
print(len(single_name_singers))    # length of a list
print(sorted(single_name_singers)) # sort a list
print(sum(primes),",",max(primes)) # sum and max of a list

6
['Beyoncé', 'Eminem', 'Madonna', 'P!nk', 'Prince', 'Rhianna']
58 , 17


There are also special functions "attached" to any list. Functions like this are called _methods_. (This is [object oriented programming](https://realpython.com/python3-object-oriented-programming/) termonology - but we don't need the details for now.) 

For example, we find the index for a given value using the `index` method

In [48]:
i = primes.index(7)
print(i,primes[i]==7)

3 True


Which shows that 7 is at index 3, agreeing with our slicing above.

Note that the format of calling a method is different than a regular function. We write the object, then a dot, then the name of the method, then the arguments in parethesis. In this example `prime` is the object, `index` is the method, and `7` was the argument. The `format` statement is also a method. In the code `"I am {} years old".format(19)`, the string is the object, `format` is the method and `19` is the argument.

## Creating lists

We've seen that we can write down a list inside square brackets, but this isn't very practical for long lists. If the list is just a simple repetition, you can create it using the multiply operation:

In [49]:
a = [4]*5
b = [5]*4
print(a,b)

[4, 4, 4, 4, 4] [5, 5, 5, 5]


And we can append lists together using the addition operation:

In [50]:
print(a+[8])
print(a+b+a)

[4, 4, 4, 4, 4, 8]
[4, 4, 4, 4, 4, 5, 5, 5, 5, 4, 4, 4, 4, 4]


This is called operator overloading. Python decides what the `*,+` operations mean depending on the data type.

However, this kind of code can be confusing and still isn't very general. The best way to create lists is using a [list comprehension](https://docs.python.org/3/tutorial/datastructures.html?highlight=comprehension#list-comprehensions). Basically, we loop through a simple list like `range(n)` to create a new list inside the square brackets.

Now instead of printing our wavelengths to the screen, let's make a list of them, so we can use them for other parts of our program later.

In [51]:
omega_list = range(1,20) # simple list using range
wavelength_list = [deep_wavelength(omega) for omega in omega_list] # list comprehension
wavelength_list

[61.6379958,
 15.40949895,
 6.8486662,
 3.8523747375,
 2.465519832,
 1.71216655,
 1.257918281632653,
 0.963093684375,
 0.7609629111111111,
 0.616379958,
 0.5094049239669421,
 0.4280416375,
 0.36472186863905326,
 0.31447957040816327,
 0.273946648,
 0.24077342109375,
 0.21328026228373703,
 0.19024072777777778,
 0.1707423706371191]

Note that you *could* make the second list with a for loop

```python
wavelength_list=[]
for omega in omega_list:
    wavelength_list = wavelength_list + [deep_wavelength(omega)]
```

**But don't!** A list comprehension is computationally faster and it is more readable.

Let's also make a list of the frequencies in cycles per second.

In [52]:
freq_list = [omega/(2*3.14159) for omega in omega_list]
freq_list

[0.15915507752443828,
 0.31831015504887655,
 0.4774652325733148,
 0.6366203100977531,
 0.7957753876221914,
 0.9549304651466296,
 1.1140855426710679,
 1.2732406201955062,
 1.4323956977199443,
 1.5915507752443827,
 1.7507058527688208,
 1.9098609302932592,
 2.0690160078176976,
 2.2281710853421357,
 2.387326162866574,
 2.5464812403910124,
 2.7056363179154506,
 2.8647913954398887,
 3.0239464729643273]

Finally, let's join these two equal length lists "side-by-side" using `zip`. This lets me loop through them simultaneously and print both the frequency and wavelength, but only for waves that are short enough to be valid "deep water" waves and for $f<2\text{Hz}$ to avoid overstressing the wavemaker hydraulics.

In [53]:
for freq,wavelength in zip(freq_list,wavelength_list):
    if(wavelength<6 and freq<2):
        print("{:.3g} Hz| {:.3g} m".format(freq,wavelength))

0.637 Hz| 3.85 m
0.796 Hz| 2.47 m
0.955 Hz| 1.71 m
1.11 Hz| 1.26 m
1.27 Hz| 0.963 m
1.43 Hz| 0.761 m
1.59 Hz| 0.616 m
1.75 Hz| 0.509 m
1.91 Hz| 0.428 m


## Example exercise

Create a list with the wave speed $c=\lambda f$ and the time it would take for each wave to reach the middle of the $130\text{m}$ long tank. You can make a really big breaking wave by timing it such that all of these individual waves meet-up at center at the same time. When do you need to generate each wave to achieve this? Limit your answers to the waves shorter than $6m$ and less than $2Hz$.

To help you check your answers, the slowest wave has the following metrics:

| freq $(Hz)$ | speed $(m/s)$ | time $(s)$ | delay $(s)$ |
|--|--|--|--|--|
| 1.91 | 0.818 | 79.5 | 0 |

In [77]:
flist = []
lamlist = []

for freq,wavelength in zip(freq_list,wavelength_list):
    if(wavelength<6 and freq<2):
        flist.append(freq)
        lamlist.append(wavelength)


# speed_list = []

speed_list = [f * lam for f, lam in zip(flist, lamlist)]

# time_list = []

time_list = [130/ 2 /speed for speed in speed_list]

# delay_list = []

delay_list = [max(time_list) - time for time in time_list]
delay_list

[53.00713557594292,
 46.38124362895005,
 39.75535168195719,
 33.12945973496432,
 26.503567787971463,
 19.877675840978597,
 13.251783893985731,
 6.625891946992866,
 0.0]